Clean up API

towerofnix opened the pull request:
Fixes #196 by changing the API as described there. This is obviously a breaking change (the client is updated in this PR as well).

> This first pass focuses on two things:
>
> 1. Destroy any usage of `{success: true}`.
> 2. Make sure all returned values are an object, with specific details (e.g. user list) responded in separate properties (e.g. `"users": []`) and not stored on the top-level returned object itself.
>
> ---
>
> This part finishes up the clean-up, adding proper errors. Docs need to be updated.

heyitsmeuralex commented on docs/api.md:
```diff
@@ -193,28 +193,48 @@ Returns `{available}`, where `available` is a boolean set to whether or not the
+
+## Endpoint errors
+
+Nearly all of the [HTTP endpoints](#http-endpoints) return errors situationally. The following list describes each:
```
Could you describe the `{ error: { code, message }}` format here?

heyitsmeuralex commented on packages/server/api.js:
```diff
@@ -487,7 +490,7 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
             // the username variable is passed. To make this a little less
             // evil, it's possible for that word to be passed manually
             // (as the second argument to requireNameValid).
```
Does this comment still apply?

towerofnix replied:
Nope, thanks. Removed (commit to be pushed).

Funnily enough, I don't think anything actually passed an `errorFieldName` to `requireNameValid`.

heyitsmeuralex commented on packages/server/api.js:
```diff
@@ -661,12 +664,12 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
       if (err) {
         res.status(500).end(JSON.stringify({
-          error: err.message
+          error: Object.assign(errors.UPLOAD_FAILED, {message: err.message})
```
`Object.assign` mutates its first argument, doesn't it? Fix pls.

towerofnix replied:
Ooops! I meant to do `Object.assign({}, errors.UPLOAD_FAILED, {message: err.message})`. Thanks.

heyitsmeuralex commented on packages/server/api.js:
```diff
@@ -1384,7 +1360,7 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
       if (sessionUser._id === userID) {
         response.status(400).end(JSON.stringify({
-          error: 'you cannot deauthorize yourself'
+          error: errors.AUTHORIZATION_ERROR
```
Ehh... this is a mix between `NOT_YOURS` and `AUTHORIZATION_ERROR`, isn't it? Could you use a custom `message` here to clarify?

towerofnix replied:
Sure, I'll add a custom message. I didn't put too much thought into AUTHORIZATION_ERRORs because they're going to be obsolete pretty near in the future.

heyitsmeuralex commented on packages/server/api.js:
```diff
@@ -1411,20 +1385,18 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
       } else if (sessionIDs.find(x => typeof x !== 'string')) {
         respones.status(400).end(JSON.stringify({
-          error: 'expected sessionIDs to be an array of strings'
+          error: errors.INVALID_PARAMETER_TYPE
```
Clarification of the expected type here too would be nice. We could have a `error.invalidParameterType(key, expected, actual)` for generating these..?

towerofnix replied:
I totally want an `invalidParameterType` function, but I think it's a liiiittle out of the scope of this PR.

Anyways, I'll add custom messages to these two.

heyitsmeuralex commented:
Also note merge conflicts - there's a new error in there we need to add, notably `UPLOADS_DISABLED`.

heyitsmeuralex commented:
Also it'd be nice to catch internal errors and return code 500 and `{ error: { code: 'INTERNAL_ERROR', message: 'ReferenceError: ...' }}` or something.

towerofnix commented:
>Also note merge conflicts - there's a new error in there we need to add, notably UPLOADS_DISABLED.

Right, fixed. Also fixed the other things you mentioned and added info regarding what errors actually look like to the docs.
